### ISSUE TITLE
ccache creates non existent directories

### ISSUE THREAD

**Issue by ggrandou** [ccache/ccache#353]
Hi,

I have found a case where a gcc commands hits in ccache and creates a non-existent destination directory.

Here is a small test case which shows the issue:
[test_ccache.tar.gz](https://github.com/ccache/ccache/files/2853083/test_ccache.tar.gz)

running commands in do_test gives:
```
+ which ccache
/usr/bin/ccache
+ ccache --version
ccache version 3.3.4

Copyright (C) 2002-2007 Andrew Tridgell
Copyright (C) 2009-2017 Joel Rosdahl

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; either version 3 of the License, or (at your option) any later
version.
+ which gcc
/usr/lib/ccache/gcc
++ which gcc
+ realpath /usr/lib/ccache/gcc
/usr/bin/ccache
+ gcc --version
gcc (Debian 6.3.0-18+deb9u1) 6.3.0 20170516
Copyright (C) 2016 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

+ ccache -C
Cleared cache
+ echo

+ make clean
rm -rf hello objs
+ echo

+ make hello
gcc -c -o objs/hello.o hello.c
Assembler messages:
Fatal error: can't create objs/hello.o: No such file or directory
Makefile:10: recipe for target 'objs/hello.o' failed
make: *** [objs/hello.o] Error 1
+ echo

+ make
mkdir -p objs
gcc -c -o objs/hello.o hello.c
gcc -o hello objs/hello.o
+ echo

+ make clean
rm -rf hello objs
+ echo

+ make hello
gcc -c -o objs/hello.o hello.c
gcc -o hello objs/hello.o
```

the 1st 'make hello' fails because objs/ directory does not exist
the make call creates the directory and build objs/hello.o and hello.
ccache now contains objs/hello.o

after having deleted objs/ a new call to make hello suddenly suceeds, even though the objs does not exist: the ccache call has created objs/ and objs/hello.o.

I have tested this with:
- ccache 3.3.4 from my debian stretch machine
- ccache 3.6 compiled from sources 
- run without and with CCACHE_NODIRECT env var

Behavior is always the same.

Is it a known issue ? or is it even the desired behavior ?

**Comment by jrosdahl** [comment]
> Is it a known issue ? or is it even the desired behavior ?

No, it's not a known issue. Even though it doesn't feel very important to fix, I think that it indeed would be better to mimic the compiler behavior in this case.

The behavior is an unintended side effect of refactoring in fc61ca9a9232f26ee3f714d4b8738d916f6948ff (first included in ccache 3.2).

**Comment by ggrandou** [comment]
> I think that it indeed would be better to mimic the compiler behavior in this case.

I had a bug in my project's makefile which remained unoticed until I published it and other people started to build without using ccache (or with an empty cache). So yes, that would be better ;)

Thanks for the prompt answer!
